File: apps/scanner/app/src/main/java/id/autotoko/scanner/Units.java
package id.autotoko.scanner;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

/**
 * Turning what the box says into what the catalogue counts.
 *
 * A supplier ships glycerine in a 1 kg jug and a 5 kg jug; the catalogue holds
 * glycerine in grams, because grams is what a recipe consumes. Somebody has to
 * bridge that, and until now it was the packer, in their head, into a field
 * labelled only "gram". Typing 1 for a 1 kg jug is not carelessness — it is the
 * obvious reading of the label — and it credits the shelf with a thousandth of
 * what arrived, with nothing anywhere looking wrong afterwards.
 *
 * Mirrors packages/shared/src/units.ts. Kept as a copy rather than fetched,
 * because the sheet has to convert while the packer types and the phone is
 * routinely on a warehouse connection that a lookup cannot depend on.
 *
 * No density table on purpose: millilitres of an oil are not grams of it, and
 * a 1.0 that is right for water is wrong for everything anyone actually buys.
 */
public final class Units {

    public static final String MASS = "mass";
    public static final String VOLUME = "volume";
    public static final String COUNT = "count";
    public static final String UNKNOWN = "unknown";

    private static final class Def {
        final String kind;
        final double factor;
        final String label;

        Def(String kind, double factor, String label) {
            this.kind = kind;
            this.factor = factor;
            this.label = label;
        }
    }

    private static final Map<String, Def> UNITS = new HashMap<>();

    static {
        // mass, base = gram
        put("mg", MASS, 0.001, "mg");
        put("miligram", MASS, 0.001, "mg");
        put("g", MASS, 1, "gram");
        put("gr", MASS, 1, "gram");
        put("gram", MASS, 1, "gram");
        put("grams", MASS, 1, "gram");
        put("ons", MASS, 100, "ons");
        put("kg", MASS, 1000, "kg");
        put("kilo", MASS, 1000, "kg");
        put("kilogram", MASS, 1000, "kg");
        put("ton", MASS, 1000000, "ton");

        // volume, base = ml
        put("ml", VOLUME, 1, "ml");
        put("mililiter", VOLUME, 1, "ml");
        put("milliliter", VOLUME, 1, "ml");
        put("cc", VOLUME, 1, "ml");
        put("l", VOLUME, 1000, "liter");
        put("lt", VOLUME, 1000, "liter");
        put("ltr", VOLUME, 1000, "liter");
        put("liter", VOLUME, 1000, "liter");
        put("litre", VOLUME, 1000, "liter");
        put("galon", VOLUME, 19000, "galon");

        // count, base = one thing
        put("pcs", COUNT, 1, "pcs");
        put("pc", COUNT, 1, "pcs");
        put("buah", COUNT, 1, "pcs");
        put("unit", COUNT, 1, "pcs");
        put("biji", COUNT, 1, "pcs");
        put("lembar", COUNT, 1, "lembar");
        put("sheet", COUNT, 1, "lembar");
        put("label", COUNT, 1, "label");
        put("roll", COUNT, 1, "roll");
        put("botol", COUNT, 1, "botol");
        put("lusin", COUNT, 12, "lusin");
        put("dus", COUNT, 1, "dus");
        put("box", COUNT, 1, "box");
    }

    private static void put(String key, String kind, double factor, String label) {
        UNITS.put(key, new Def(kind, factor, label));
    }

    private Units() {}

    public static String normalize(String unit) {
        if (unit == null) return "";
        String u = unit.trim().toLowerCase(Locale.ROOT);
        while (u.endsWith(".") || u.endsWith(" ")) u = u.substring(0, u.length() - 1);
        return u;
    }

    public static String kindOf(String unit) {
        Def d = UNITS.get(normalize(unit));
        return d == null ? UNKNOWN : d.kind;
    }

    /**
     * What the packer may sensibly choose for a material held in `target`.
     *
     * The catalogue's own unit comes first, because most of the time the label
     * really does agree with it and that should not be the awkward pick.
     */
    public static List<String> compatible(String target) {
        String t = normalize(target);
        Def def = UNITS.get(t);
        Set<String> out = new LinkedHashSet<>();
        if (def == null) {
            if (!t.isEmpty()) out.add(t);
            return new ArrayList<>(out);
        }
        out.add(def.label);
        if (COUNT.equals(def.kind)) {
            // A count is only ever itself. "lusin" is twelve "pcs", but a
            // "botol" is not a "roll" and offering them together invites a
            // wrong pick that nothing downstream would catch.
            if ("pcs".equals(def.label) || "lusin".equals(def.label)) {
                out.add("pcs");
                out.add("lusin");
            }
            return new ArrayList<>(out);
        }
        for (Def d : UNITS.values()) {
            if (d.kind.equals(def.kind)) out.add(d.label);
        }
        return new ArrayList<>(out);
    }

    /**
     * `value` of unit `from`, expressed in `to`; null when it cannot be done.
     *
     * Null rather than a plausible number. Grams asked to become millilitres
     * has no answer without a density, and returning one anyway would put a
     * quantity of nothing on the shelf.
     */
    public static Double convert(double value, String from, String to) {
        if (Double.isNaN(value) || Double.isInfinite(value)) return null;

        String f = normalize(from);
        String t = normalize(to);

        // Same unit, or the catalogue never stated one: nothing to convert,
        // and inventing a conversion for a blank is worse than passing through.
        if (f.equals(t)) return value;
        if (f.isEmpty() || t.isEmpty()) return value;

        Def a = UNITS.get(f);
        Def b = UNITS.get(t);
        if (a == null || b == null) return null;
        if (!a.kind.equals(b.kind)) return null;

        return (value * a.factor) / b.factor;
    }

    /** "6.000 gram" — what the shelf will actually receive, spelled out. */
    public static String describe(Double qty, String unit) {
        if (qty == null) return "-";
        String n = qty == Math.floor(qty) && !Double.isInfinite(qty)
                ? String.format(Locale.GERMAN, "%,.0f", qty)
                : String.format(Locale.GERMAN, "%,.3f", qty);
        return unit == null || unit.isEmpty() ? n : n + " " + unit;
    }
}

File: packages/shared/src/units.ts
/**
 * Converting what arrived into what the recipe counts in.
 *
 * A supplier ships glycerine in a 1 kg jug and a 5 kg jug; the catalogue holds
 * glycerine in grams because that is what a recipe consumes. Somebody has to
 * turn one into the other, and until now that somebody was the packer, in their
 * head, into a box labelled only "gram". Typing 1 instead of 1000 there is not
 * an unlikely mistake — it is the obvious reading of "1 kg" — and it understates
 * the shelf by a factor of a thousand without anything looking wrong.
 *
 * Only mass and volume convert. Deliberately no density: millilitres of an oil
 * are not grams of it, and a table that quietly picked 1.0 would be wrong for
 * every material anyone actually buys.
 */

/** What a unit measures. Units of different kinds never convert. */
export type UnitKind = "mass" | "volume" | "count" | "unknown";

interface UnitDef {
  kind: UnitKind;
  /** How many base units (gram for mass, ml for volume) one of these is. */
  factor: number;
  /** What to show in a picker. */
  label: string;
}

/**
 * Spelling is not the packer's problem. The catalogue already holds "Pcs" and
 * "pcs" as separate strings, and a supplier writes "Kg", "KG" and "kilogram"
 * for the same jug, so everything is matched lowercased and trimmed.
 */
const UNITS: Record<string, UnitDef> = {
  // mass, base = gram
  mg: { kind: "mass", factor: 0.001, label: "mg" },
  miligram: { kind: "mass", factor: 0.001, label: "mg" },
  g: { kind: "mass", factor: 1, label: "gram" },
  gr: { kind: "mass", factor: 1, label: "gram" },
  gram: { kind: "mass", factor: 1, label: "gram" },
  grams: { kind: "mass", factor: 1, label: "gram" },
  ons: { kind: "mass", factor: 100, label: "ons" },
  kg: { kind: "mass", factor: 1000, label: "kg" },
  kilo: { kind: "mass", factor: 1000, label: "kg" },
  kilogram: { kind: "mass", factor: 1000, label: "kg" },
  ton: { kind: "mass", factor: 1_000_000, label: "ton" },

  // volume, base = ml
  ml: { kind: "volume", factor: 1, label: "ml" },
  mililiter: { kind: "volume", factor: 1, label: "ml" },
  milliliter: { kind: "volume", factor: 1, label: "ml" },
  cc: { kind: "volume", factor: 1, label: "ml" },
  l: { kind: "volume", factor: 1000, label: "liter" },
  lt: { kind: "volume", factor: 1000, label: "liter" },
  ltr: { kind: "volume", factor: 1000, label: "liter" },
  liter: { kind: "volume", factor: 1000, label: "liter" },
  litre: { kind: "volume", factor: 1000, label: "liter" },
  galon: { kind: "volume", factor: 19_000, label: "galon" },

  // count, base = one thing
  pcs: { kind: "count", factor: 1, label: "pcs" },
  pc: { kind: "count", factor: 1, label: "pcs" },
  buah: { kind: "count", factor: 1, label: "pcs" },
  unit: { kind: "count", factor: 1, label: "pcs" },
  biji: { kind: "count", factor: 1, label: "pcs" },
  lembar: { kind: "count", factor: 1, label: "lembar" },
  sheet: { kind: "count", factor: 1, label: "lembar" },
  label: { kind: "count", factor: 1, label: "label" },
  roll: { kind: "count", factor: 1, label: "roll" },
  botol: { kind: "count", factor: 1, label: "botol" },
  lusin: { kind: "count", factor: 12, label: "lusin" },
  dus: { kind: "count", factor: 1, label: "dus" },
  box: { kind: "count", factor: 1, label: "box" },
};

export function normalizeUnit(unit: string | null | undefined): string {
  return (unit ?? "").trim().toLowerCase().replace(/[.\s]+$/g, "");
}

export function unitKind(unit: string | null | undefined): UnitKind {
  const def = UNITS[normalizeUnit(unit)];
  return def ? def.kind : "unknown";
}

/**
 * Units the packer may sensibly enter for a material held in `target`.
 *
 * Ordered with the catalogue's own unit first, because most of the time the
 * label really does say what the catalogue says and the picker should not make
 * that the awkward choice.
 */
export function compatibleUnits(target: string | null | undefined): string[] {
  const t = normalizeUnit(target);
  const def = UNITS[t];
  if (!def) return t ? [t] : [];

  const seen = new Set<string>();
  const out: string[] = [];
  const push = (label: string) => {
    if (!seen.has(label)) {
      seen.add(label);
      out.push(label);
    }
  };
  push(def.label);
  // A count is only ever itself: "lusin" converts to "pcs", but "botol" and
  // "roll" are not each other and offering them together invites a wrong pick.
  if (def.kind === "count") {
    if (def.label === "pcs" || def.label === "lusin") {
      push("pcs");
      push("lusin");
    }
    return out;
  }
  for (const d of Object.values(UNITS)) {
    if (d.kind === def.kind) push(d.label);
  }
  return out;
}

/**
 * `value` of unit `from`, expressed in unit `to`.
 *
 * Returns null when it cannot be done rather than a number that looks fine:
 * an unrecognised unit, or grams asked to become millilitres. The caller has to
 * decide what to do about that, and every caller here refuses the entry.
 */
export function convertUnit(
  value: number,
  from: string | null | undefined,
  to: string | null | undefined,
): number | null {
  if (!Number.isFinite(value)) return null;

  const f = normalizeUnit(from);
  const t = normalizeUnit(to);

  // Same spelling, or the catalogue never stated a unit: nothing to convert,
  // and inventing a conversion for a blank unit would be worse than leaving
  // the number as the packer typed it.
  if (f === t) return value;
  if (!f || !t) return value;

  const a = UNITS[f];
  const b = UNITS[t];
  if (!a || !b) return null;
  if (a.kind !== b.kind) return null;

  return (value * a.factor) / b.factor;
}

/** True when the two units describe the same kind of thing. */
export function unitsCompatible(a: string | null | undefined, b: string | null | undefined): boolean {
  const x = normalizeUnit(a);
  const y = normalizeUnit(b);
  if (x === y) return true;
  if (!x || !y) return true;
  const da = UNITS[x];
  const db = UNITS[y];
  if (!da || !db) return false;
  return da.kind === db.kind;
}
